upstairs/src/impacted_blocks.rs

```diff
@@ -0,0 +1,275 @@
+// Copyright 2022 Oxide Computer Company
+
+use super::*;
+
+use bitvec::prelude::*;
+use itertools::Itertools;
+use std::ops::BitAnd;
+
+/// Store a list of impacted blocks for later job dependency calculation
+#[derive(Debug, Default, PartialEq)]
+pub struct ImpactedBlocks {
+    ddef: RegionDefinition,
+
+    // extent id -> list of block offsets
+    blocks: HashMap<u64, BitVec>,
+}
+
+impl ImpactedBlocks {
+    pub fn new(ddef: RegionDefinition) -> Self {
+        ImpactedBlocks {
+            ddef,
+            blocks: HashMap::default(),
+        }
+    }
+
+    pub fn blocks_in_extent(&self) -> usize {
+        self.ddef.extent_size().value as usize
+    }
+
+    pub fn add(&mut self, extent_id: u64, block: Block) {
+        let blocks_in_extent = self.blocks_in_extent();
+        let bv = self.blocks.entry(extent_id).or_insert_with(|| {
+            let mut bv = BitVec::with_capacity(blocks_in_extent);
+            bv.resize(blocks_in_extent, false);
+            bv
+        });
+
+        bv.set(block.value as usize, true);
+    }
+
+    /// Return true if this list of impacted blocks overlaps with another.
+    pub fn conflicts(&self, other: &ImpactedBlocks) -> bool {
+        for shared_key in self
+            .blocks
+            .keys()
+            .filter(|key| other.blocks.contains_key(key))
+        {
+            // TODO any way to avoid this clone?
+            let bv = self.blocks[shared_key].clone();
+            if bv.bitand(&other.blocks[shared_key]).any() {
+                return true;
+            }
+        }
+
+        false
+    }
+
+    /// Returns (extent id, block) tuples
+    pub fn tuples(&self) -> Vec<(u64, Block)> {
+        let mut result = Vec::with_capacity(self.len());
+
+        let sorted_keys = self.blocks.keys().sorted();
+        for eid in sorted_keys {
+            let block_offsets =
+                self.blocks[eid].iter_ones().collect::<Vec<usize>>();
+            for i in block_offsets.iter().sorted() {
+                result
+                    .push((*eid, Block::new_with_ddef(*i as u64, &self.ddef)));
+            }
+        }
+
+        result
+    }
+
+    /// Returns the number of impacted blocks
+    pub fn len(&self) -> usize {
+        let mut len = 0;
+
+        for bv in self.blocks.values() {
+            len += bv.iter_ones().count();
+        }
+
+        len
+    }
+
+    #[allow(dead_code)]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+}
+
+/// Given an offset and number of blocks, return a list of individually impacted
+/// blocks:
```

Comment:
This returns a list (hashmap) of extents, and the impacted blocks for that extent, right?

Reply:
Yeah, `ImpactedBlocks` has

```rust
    // extent id -> list of block offsets
    blocks: HashMap<u64, BitVec>,
```

Reply:
improved comment in 2ec554f